src/Microsoft.AspNet.Mvc.DataAnnotations/CompareAttributeAdapter.cs

```diff
@@ -28,12 +28,17 @@ public CompareAttributeAdapter(CompareAttribute attribute, IStringLocalizer stri
                 throw new ArgumentNullException(nameof(context));
             }
 
-            var errorMessage = ((CompareAttributeWrapper)Attribute).FormatErrorMessage(context);
+            var errorMessage = GetErrorMessage(context.ModelMetadata, context.MetadataProvider);
```

Comment:
Why is this class not more like the other adapter classes? Unclear `CompareAttributeWrapper` is still needed with your changes in place.